src/library/scala/collection/mutable/HashSet.scala

```diff
@@ -28,64 +29,213 @@ package mutable
   * @define mayNotTerminateInf
   * @define willNotTerminateInf
   */
-final class HashSet[A]
+final class HashSet[A](initialCapacity: Int, loadFactor: Double)
   extends AbstractSet[A]
     with SetOps[A, HashSet, HashSet[A]]
     with StrictOptimizedIterableOps[A, HashSet, HashSet[A]] {
 
-  @transient private[this] var table = new FlatHashTable[A]
+  def this() = this(HashSet.defaultInitialCapacity, HashSet.defaultLoadFactor)
 
-  // Used by scala-java8-compat (private[mutable] erases to public, so Java code can access it)
-  private[mutable] def getTable: FlatHashTable[A] = table
-  override def iterator: Iterator[A] = table.iterator
+  import HashSet.Node
 
-  override def iterableFactory: IterableFactory[HashSet] = HashSet
+  /** The actual hash table. */
+  private[this] var table = new Array[Node[A]](tableSizeFor(initialCapacity))
+
+  /** The next size value at which to resize (capacity * load factor). */
+  private[this] var threshold: Int = newThreshold(table.length)
+
+  private[this] var contentSize = 0
+
+  override def size: Int = contentSize
 
-  def addOne(elem: A): this.type = {
-    table.addElem(elem)
-    this
+  @`inline` private[this] final def computeHash(o: A): Int = {
+    val h = o.##
+    h ^ (h >>> 16)
   }
-  def subtractOne(elem: A): this.type = {
-    table.removeElem(elem)
-    this
+
+  @`inline` private[this] final def index(hash: Int) = hash & (table.length - 1)
+
+  override def contains(elem: A): Boolean = findNode(elem) ne null
+
+  @`inline` private[this] def findNode(elem: A): Node[A] = {
+    val hash = computeHash(elem)
+    table(index(hash)) match {
+      case null => null
+      case nd => nd.findNode(elem, hash)
+    }
+  }
+
+  override def sizeHint(size: Int): Unit = {
+    val target = tableSizeFor(((size + 1).toDouble / loadFactor).toInt)
+    if(target > table.length) growTable(target)
   }
 
-  override def add(elem: A): Boolean = table.addElem(elem)
+  override def add(elem: A) : Boolean = {
+    if(contentSize + 1 >= threshold) growTable(table.length * 2)
+    addElem(elem, computeHash(elem))
+  }
+
+  override def addAll(xs: IterableOnce[A]): this.type = {
+    sizeHint(xs.knownSize)
+    super.addAll(xs)
+  }
 
-  override def remove(elem: A): Boolean = table.removeElem(elem)
+  private[this] def addElem(elem: A, hash: Int) : Boolean = {
+    val idx = index(hash)
+    table(idx) match {
+      case null =>
+        table(idx) = new Node(elem, hash, null)
+      case old =>
+        var prev: Node[A] = null
+        var n = old
+        while((n ne null) && n.hash <= hash) {
+          if(n.hash == hash && elem == old.key) return false
```

Comment:
Here's the bug that causes the compiler crashes. This needs to be `n.key` instead of `old.key`.